### PR TITLE
Split data generator

### PR DESCRIPTION
**Pull request by DeNeutoy**

### CODE REVIEW

**Comment by DeNeutoy** [comment]
@matt-gardner Do you mind modifying the existing data generator tests to test this? I don't exactly understand how the `FakeEncoders` etc work in that test and it's quite complicated. I'm relying on this API for #14.

**Comment by matt-gardner** [comment]
Yes, I can handle tests for this.  I should have a solid block of time tomorrow to just focus on code and get a bunch of things done that have been sitting for too long...

File: allennlp/data/iterators/bucket_iterator.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+        self.sorting_keys = sorting_keys or []
+        self.padding_noise = padding_noise
+        self.sort_every_epoch = sort_every_epoch
+        super(BucketIterator, self).__init__(batch_size)
+
+    @overrides
+    def __call__(self, dataset: Dataset):
+        grouped_instances = self._create_batches(dataset)
+        self.last_num_batches = len(grouped_instances)  # pylint: disable=attribute-defined-outside-init
```
Comment (matt-peters) [inline allennlp/data/iterators/bucket_iterator.py]: Where is this attribute used?  It seems a little scary to assign something to `self` in `__call__` since a client has to remember the last time they called the iterator to know what the value of this attribute is?

Comment (matt-peters) [inline allennlp/data/iterators/bucket_iterator.py]: Alternatively, if properties of the `Dataset` are needed after calling the iterator then it makes more sense to change this to an `__iter__` API (see below), and pass the `Dataset` into the constructor.  Then each instance of a `BucketIterator` would be associated with one `Dataset`, and could store attributes that depend on both the dataset and the iterator like the number of batches, etc.

Comment (matt-gardner) [inline allennlp/data/iterators/bucket_iterator.py]: Yes, this was an unfortunate affect of how the previous API handled things, where we tried to extend the API after the fact, and resorted to scary state.  If we can find a better way to handle this in the redesign, that'd be better.

The problem was that we didn't want to have to change the return value of `as_training_data()`, but need to know how many batches there were in the dataset, to define an "epoch" for `fit_generator` in Keras.  If we design this correctly, we should be able to avoid this issue entirely.

Comment (DeNeutoy) [inline allennlp/data/iterators/bucket_iterator.py]: I would vote for removing the idea of an "epoch" entirely and just verifying training in terms of number of gradient updates.

Comment (matt-peters) [inline allennlp/data/iterators/bucket_iterator.py]: I still think it's important to have a notion of an "epoch" or a batch size independent measure of how much data we have iterated through (especially for testing where we want to iterate for one epoch exactly).

File: allennlp/data/iterators/bucket_iterator.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+        while True:
+            if self.sort_every_epoch:
+                unpadded_dataset = deepcopy(dataset)
+                groups = self._create_batches(unpadded_dataset)
+            else:
+                groups = grouped_instances
+            for group in groups:
+                batch = Dataset(group)
```
Comment (matt-gardner) [inline allennlp/data/iterators/bucket_iterator.py]: I'd expect this overhead to be negligible relative to training time.

File: allennlp/data/iterators/iterator.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+from ..dataset import Dataset
+from ...common.params import Params
+
+
+class Iterator:
+    """
+    An abstract Iterator class. Iterators must implement __call__, which yields
+    batched examples.
+    """
+    def __call__(self, dataset: Dataset) -> Generator[Dict[str, List[numpy.array]], None, None]:
```
Comment (matt-gardner) [inline allennlp/data/iterators/iterator.py]: I talked with Mark about this, and I think `__call__` is more flexible here - it can take parameters, and can be changed and added to later, while `__iter__` locks you into a particular way of doing things that can't be changed easily.  Also, doing it this way lets us only construct the iterator once, with whatever state is necessary (e.g., batch size, dynamic padding, adaptive batch size...), and use it for all datasets (training, validation, whatever).  The reason we had the ugly `last_num_batches` thing was to ensure that we could keep track of an epoch, and only iterate over a dataset exactly once.  I just added a `yield_one_pass` function to do that instead, as well as a `num_batches_per_epoch` function, if you don't want to use `yield_one_pass`.

**Comment by matt-gardner** [comment]
@DeNeutoy, I'm still adding a few more tests, but the bulk code changes are ready to be looked at.

File: allennlp/data/iterators/iterator.py (hunk 3 above)
Comment (matt-peters) [inline allennlp/data/iterators/iterator.py]: Those are definitely good reasons to prefer `__call__` over `__iter__`, it's more flexible.

File: allennlp/data/iterators/data_iterator.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        """
+        Returns a generator that yields batches over the given dataset, forever.
+
+        Parameters
+        ----------
+        dataset : ``Dataset``
+        """
+        raise NotImplementedError
+
+    def yield_one_pass(self, dataset: Dataset) -> Generator[Dict[str, List[numpy.array]], None, None]:
```
Comment (matt-peters) [inline allennlp/data/iterators/data_iterator.py]: Can we streamline the API a little here?  `yield_one_pass` is essentially the same thing as `__call__`, just with a specific termination point and it seems strange to me that it has a completely different name and calling convention.  Since `__call__` can take parameters, how about just adding a `n_epochs=None` parameter to `__call__` that defaults to iterating forever?

File: allennlp/data/iterators/basic_iterator.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+    @overrides
+    def yield_one_pass(self, dataset: Dataset):
+        grouped_instances = self._create_batches(dataset)
+        for group in grouped_instances:
+            batch = Dataset(group)
+            padding_lengths = batch.get_padding_lengths()
+            yield batch.as_arrays(padding_lengths, verbose=False)
+
+    def _create_batches(self, dataset: Dataset, shuffle: bool = True) -> List[List[Instance]]:
```
Comment (matt-peters) [inline allennlp/data/iterators/basic_iterator.py]: Since `_create_batches` is only used internally, I don't see how a client can set the `shuffle` parameter here?  It needs to be passed down through `__call__` (or `yield_one_pass`) or specified in the constructor.

File: allennlp/data/fields/text_field.py [hunk 6]
```diff
@@ -42,6 +42,7 @@ def count_vocab_items(self, counter: Dict[str, Dict[str, int]]):
     def index(self, vocab: Vocabulary):
         token_arrays = []
         for indexer in self._token_indexers:
+            print("Indexer:", indexer)
```
Comment (DeNeutoy) [inline allennlp/data/fields/text_field.py]: remove print statement.

File: allennlp/data/instance.py [hunk 7]
```diff
@@ -60,3 +60,9 @@ def pad(self, padding_lengths: Dict[str, Dict[str, int]]) -> Dict[str, List[nump
 
     def fields(self):
         return self._fields
+
+    def get_field(self, name: str):
```
Comment (DeNeutoy) [inline allennlp/data/instance.py]: This unnecessarily clutters the API when we can use `self.fields()["name"]` instead. It also stops us accidentally using this in allennlp internals where we assume that an `Instance` has a key or something.

File: allennlp/data/iterators/adaptive_iterator.py [hunk 8]
```diff
@@ -39,67 +60,43 @@ class AdaptiveIterator(BucketIterator):
...
-        super(AdaptiveIterator, self).__init__(sorting_keys, padding_noise, sort_every_epoch)
+                 padding_noise: float = 0.2):
+        self._padding_memory_scaling = padding_memory_scaling
+        self._maximum_batch_size = maximum_batch_size
+        self._adaptive_memory_usage_constant = adaptive_memory_usage_constant
+        super(AdaptiveIterator, self).__init__(sorting_keys, padding_noise, biggest_batch_first)
 
     @overrides
-    def _create_batches(self, dataset: Dataset) -> List[List[Instance]]:
+    def num_batches_per_epoch(self, dataset: Dataset) -> int:
```
Comment (DeNeutoy) [inline allennlp/data/iterators/adaptive_iterator.py]: Does this concept even make sense for an adaptive iterator? If padding noise is non-zero, this is non-deterministic which is a little counter intuitive (and makes the argument for not having this as a side effect of `__call__` weaker) and having to actually process the dataset just to return this value seems a little heavy.

File: allennlp/data/iterators/basic_iterator.py (hunk 5 above)
Comment (matt-gardner) [inline allennlp/data/iterators/basic_iterator.py]: Do you want the client to be able to specify it?  I was just using it in subclasses (see `biggest_batch_first` in `BucketIterator`).

File: allennlp/data/iterators/bucket_iterator.py [hunk 9]
```diff
@@ -1,90 +1,85 @@
...
+                                                    self._sorting_keys,
+                                                    self._padding_noise)
+        grouped_instances = super(BucketIterator, self)._create_batches(dataset, shuffle=False)
+        if self._biggest_batch_first:
+            # We'll actually pop the last _two_ batches, because the last one might not be full.
+            last_batch = grouped_instances.pop()
+            penultimate_batch = grouped_instances.pop()
+        if shuffle:
+            random.shuffle(grouped_instances)
+        if self._biggest_batch_first:
```
Comment (DeNeutoy) [inline allennlp/data/iterators/bucket_iterator.py]: Yeah, this is a better place for this. Thanks for moving.

File: allennlp/data/iterators/data_iterator.py (hunk 4 above)
Comment (matt-gardner) [inline allennlp/data/iterators/data_iterator.py]: I could go either way here.  I originally had it the way you suggest, with a private `_yield_one_pass` method, then changed it to how it is now, thinking that having fewer parameters to `__call__` was probably better.  But, I can easily change it back.  @DeNeutoy, any thoughts?

**Comment by matt-peters** [comment]
Yes it definitely seems like something we should support. I can think of a few use cases for it, but mostly when a client needs to run the model on a known list of instances (so they can do further processing that requires aligning the model output with the raw instance data).

File: allennlp/data/iterators/basic_iterator.py (hunk 5 above)
Comment (matt-gardner) [inline allennlp/data/iterators/basic_iterator.py]: Ok, I'll add it.

File: allennlp/data/iterators/basic_iterator.py [hunk 10]
```diff
@@ -17,21 +20,32 @@ class BasicIterator:
...
     def __call__(self, dataset: Dataset):
-        grouped_instances = self._create_batches(dataset)
-        self.last_num_batches = len(grouped_instances)  # pylint: disable=attribute-defined-outside-init
-
-        padding_lengths = dataset.get_padding_lengths()
         while True:
-            for group in grouped_instances:
-                batch = Dataset(group)
-                yield batch.as_arrays(padding_lengths, verbose=False)
+            for batch in self.yield_one_pass(dataset):
```
Comment (DeNeutoy) [inline allennlp/data/iterators/basic_iterator.py]: you could use `yield from self.yield_one_pass(dataset)` here if we keep this way of iterating.

File: allennlp/data/iterators/data_iterator.py (hunk 4 above)
Comment (DeNeutoy) [inline allennlp/data/iterators/data_iterator.py]: Slight preference for a flag

Comment (matt-gardner) [inline allennlp/data/iterators/data_iterator.py]: Ok, will change.

File: allennlp/data/instance.py (hunk 7 above)
Comment (matt-gardner) [inline allennlp/data/instance.py]: Ok, fair point.

File: allennlp/data/iterators/adaptive_iterator.py (hunk 8 above)
Comment (matt-gardner) [inline allennlp/data/iterators/adaptive_iterator.py]: I put this in before I added the one-pass stuff.  We probably don't need this if we have a `num_epochs` flag in `__call__`.  I'll remove it.

File: allennlp/data/iterators/basic_iterator.py (hunk 10 above)
Comment (matt-gardner) [inline allennlp/data/iterators/basic_iterator.py]: Huh, I wasn't aware of the `yield from` syntax, and got sniped looking up coroutines...

**Comment by matt-gardner** [comment]
Ok, I _think_ I've addressed all of the comments, and now have tests for all of the data iterator options.

**Comment by DeNeutoy** [comment]
Just remove the `get_field` method from `Instance` and then LGTM.
